### ISSUE TITLE
Type narrowed to `never`, incorrectly

### ISSUE THREAD

**Issue by ljharb** [Microsoft/TypeScript#27103]
**TypeScript Version:**  latest

**Search Terms:** n/a

**Code**

```ts
function f(a?: {}) {
    if (!a) {
        // filters out falsy values:
        //  - `null`
        //  - `undefined`
        //  - `false`
        //  - `NaN`
        //  - `0` / `-0`
        //  - `''`
        return {};
    }

    // here, the type of `a` has been narrowed from "optional {}"
    // to just "{} minus `object`" (object-coercible, ie, anything but null/undefined, but not `object`)
    // effectively, this is `string | number | symbol | boolean | Function`

    if (typeof a !== 'object') {
        // `a` here could be the following values:
        //   - a function
        //   - `true``
        //   - a non-zero non-`NaN`` number
        //   - a non-empty string
        //   - a symbol
        return a; // however, the type inferred here is `never`
    }

    // here, `a` has a verifiably certain type of `object`.
    // because of the type guard above.
    return a; // however, the type is reported as `{}` 
}
```

**Expected behavior:** The type of `a` inside the `if` block will be either `{}`, or better, `string | number | symbol | boolean | Function`.

**Actual behavior:** The type of `a` inside the `if` block is `never`, which is flat out incorrect.

**Playground Link:** [link](http://www.typescriptlang.org/play/#src=function%20f(a%3F%3A%20%7B%7D)%20%7B%0A%20%20%20%20if%20(!a)%20%7B%0A%20%20%20%20%20%20%20%20%2F%2F%20filters%20out%20falsy%20values%3A%0A%20%20%20%20%20%20%20%20%2F%2F%20%20-%20%60null%60%0A%20%20%20%20%20%20%20%20%2F%2F%20%20-%20%60undefined%60%0A%20%20%20%20%20%20%20%20%2F%2F%20%20-%20%60false%60%0A%20%20%20%20%20%20%20%20%2F%2F%20%20-%20%60NaN%60%0A%20%20%20%20%20%20%20%20%2F%2F%20%20-%20%600%60%20%2F%20%60-0%60%0A%20%20%20%20%20%20%20%20%2F%2F%20%20-%20%60''%60%0A%20%20%20%20%20%20%20%20return%20%7B%7D%3B%0A%20%20%20%20%7D%0A%0A%20%20%20%20%2F%2F%20here%2C%20the%20type%20of%20%60a%60%20has%20been%20narrowed%20from%20%22optional%20%7B%7D%22%0A%20%20%20%20%2F%2F%20to%20just%20%22%7B%7D%20minus%20%60object%60%22%20(object-coercible%2C%20ie%2C%20anything%20but%20null%2Fundefined%2C%20but%20not%20%60object%60)%0A%20%20%20%20%2F%2F%20effectively%2C%20this%20is%20%60string%20%7C%20number%20%7C%20symbol%20%7C%20boolean%20%7C%20Function%60%0A%0A%20%20%20%20if%20(typeof%20a%20!%3D%3D%20'object')%20%7B%0A%20%20%20%20%20%20%20%20%2F%2F%20%60a%60%20here%20could%20be%20the%20following%20values%3A%0A%20%20%20%20%20%20%20%20%2F%2F%20%20%20-%20a%20function%0A%20%20%20%20%20%20%20%20%2F%2F%20%20%20-%20%60true%60%60%0A%20%20%20%20%20%20%20%20%2F%2F%20%20%20-%20a%20non-zero%20non-%60NaN%60%60%20number%0A%20%20%20%20%20%20%20%20%2F%2F%20%20%20-%20a%20non-empty%20string%0A%20%20%20%20%20%20%20%20%2F%2F%20%20%20-%20a%20symbol%0A%20%20%20%20%20%20%20%20return%20a%3B%20%2F%2F%20however%2C%20the%20type%20inferred%20here%20is%20%60never%60%0A%20%20%20%20%7D%0A%0A%20%20%20%20%2F%2F%20here%2C%20%60a%60%20has%20a%20verifiably%20certain%20type%20of%20%60object%60.%0A%20%20%20%20%2F%2F%20because%20of%20the%20type%20guard%20above.%0A%20%20%20%20return%20a%3B%20%2F%2F%20however%2C%20the%20type%20is%20reported%20as%20%60%7B%7D%60%20%0A%7D)

**Related Issues:** not sure how to search for it

**Comment by ljharb** [comment]
This may be addressed by #27157; it can be closed or left open til the next release, as you like.
